docs/melange-for-x-developers.md

```diff
@@ -0,0 +1,601 @@
+
+<table>
+  <thead>
+    <tr>
+      <th>JavaScript</th>
+      <th>OCaml (Reason syntax)</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>
+  <pre><code>let res = (function() {
+  const x = 23;
+  const y = 34;
+  return x + y;
+})();</code></pre>
+      </td>
+      <td>
+  <pre><code>let res = {
+  let x = 23;
+  let y = 34;
+  x + y
+};</code></pre>
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+### Comments
+
+| JavaScript        | OCaml (Reason syntax) |
+| ----------------- | --------------------- |
+| `/* Comment */`   | Same                  |
+| `// Line comment` | Same                  |
+
+## For TypeScript developers
+
+The approach to typing applications using Melange differs somewhat from
+TypeScript. TypeScript has been designed with a focus on compatibility with
+JavaScript, as outlined in its [design
+goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals). On
+the other hand, Melange is built upon OCaml, a compiler known for its emphasis
+on expressiveness and safety.
+
+These are some of the differences between both.
+
+### Type inference
+
+In TypeScript, the types for the input parameters have to be defined:
+
+```javascript
+let sum = (a: number, b: number) => a + b;
+```
+
+OCaml can infer types without barely any type annotations. For example, we can
+define a function that adds two numbers as:
+
+```ocaml
+let add x y = x + y
+```
+
+### Algebraic data types
+
+It is not possible to build an ADT in TypeScript the same way as in OCaml.
+[Discriminated
+unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions)
+would be the closest analog to them, with libraries like
+[ts-pattern](https://github.com/gvergnaud/ts-pattern) as an alternative to the
+lack of support for pattern matching in the language.
+
+In OCaml, [algebraic data types
+(ADTs)](https://cs3110.github.io/textbook/chapters/data/algebraic_data_types.html)
+are a commonly used functionality of the language. They allow you to build your
+own types from small blocks. And with [pattern
+matching](https://ocaml.org/docs/data-types), it is easy to access this data.
+
+### Nominal typing
+
+In TypeScript, all typing is structural. This means that it is hard sometimes to
+establish a boundary or separation between two types that have the same
+implementation. For these cases, nominal typing can be emulated using tags:
+
+```js
+type Email = string & { readonly __tag: unique symbol };
+type City = string & { readonly __tag: unique symbol };
+```
+
+In OCaml, nominal typing is fully supported. Some of the core language types
+like [records and
+variants](https://v2.ocaml.org/manual/coreexamples.html#s%3Atut-recvariants) are
+nominal. This means that even if you declare exactly the same type twice,
+functions that operate on values from one type will not be compatible with the
+other type.
+
+There is also structural typing, used for OCaml
```

Comment:
and modules.

Reply:
Not sure I understand, can you elaborate please?